File: binding-csharp/runtime/src/test/csharp/Etch/Transport/TestMessagizer.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
using System;
using Org.Apache.Etch.Bindings.Csharp.Msg;
using Org.Apache.Etch.Bindings.Csharp.Support;
using Org.Apache.Etch.Bindings.Csharp.Util;
using NUnit.Framework;

namespace Org.Apache.Etch.Bindings.Csharp.Transport
{
    [TestFixture]
    public class TestMessagizer
    {

        private const sbyte VERSION = 3; // BinaryTaggedData.VERSION;
	private const sbyte NONE = -127; // TypeCode.NONE;
	private const sbyte BYTE = -124; // TypeCode.BYTE;
        
        private readonly Who who = new MyWho();

        //static Etch.Support.DefaultValueFactory vf = new Etch.Support.DefaultValueFactory();
        public DefaultValueFactory vf;
        public MyMessageHandler mh;
        public MyPacketSource ps;
        Messagizer m;

        public TestMessagizer()
        {
            //ps = new MyPacketSource( this );
        }

        /// <summary>
        /// Call before the test fixture
        /// </summary>
        [TestFixtureSetUp]
        public void SetUpBeforeClass() 
	    {
            vf = new MyValueFactory();
            ps = new MyPacketSource();
            mh = new MyMessageHandler();
            Resources res = new Resources();
            res.Add(TransportConsts.VALUE_FACTORY, vf);
            m = new Messagizer(ps,"foo:?Messagizer.format=binary", res);
            m.SetSession(mh);
            Console.WriteLine();
            Console.Write( "TestMessagizer" );
	    }

      /*  public void SetUpBeforeEveryTest()
        {

            ps = new MyPacketSource();
            mh = new MyMessageHandler();

            m = new Messagizer( mh, vf );
            m.SetSource( ps );

        } */

        #region Test PacketHandler methods 
	    
        public class MyWho : Who
        {
            public MyWho()
            {
                //nothing
            }
        }

        //[Test]
        //public void Up() 
        //{
        //    SetUpBeforeEveryTest();

        //    PacketHandler ph = m;

        //    Assert.IsNull(mh.what);
        //    Assert.IsNull( mh.src );

        //    ph.Up( ps );

        //    Assert.AreEqual( mh.src, m );
        //    Assert.AreEqual(mh.what, What.TESTMESSAGEHANDLERUP);
        //    Assert.AreEqual( ps, m.GetPacketSource() );
        //}

        [Test]
        public void Packet1()
        {
           

        //    PacketHandler ph = m;
            SessionPacket ph = m;

            FlexBuffer buf = new FlexBuffer( new byte[] { (byte)VERSION, 
                unchecked((byte)BYTE), 1, 0, unchecked((byte)NONE) } );

            mh.handled = true;

         //   ph.Packet( who, buf );
            ph.SessionPacket(who,buf);

            Assert.AreEqual( What.TESTMESSAGEHANDLERMESSAGE, mh.what );
            Assert.AreEqual( ps, m.GetTransport() );
            Assert.AreEqual( who, mh.xsender);
            Assert.IsNotNull( mh.xmsg );
            Assert.AreEqual( MyValueFactory.mt_add, mh.xmsg.GetXType );
            Assert.AreEqual( 0, mh.xmsg.Count );

        }

        [Test]
        public void Packet2()
        {
           

            SessionPacket ph = m;

            FlexBuffer buf = new FlexBuffer(new byte[] { (byte)VERSION, 
                unchecked((byte)BYTE), 1, 0, unchecked((byte)NONE) });

            mh.handled = false;

            ph.SessionPacket( who, buf );

            Assert.AreEqual(What.OOB_NOTIFY_HANDLER, mh.what);
            Assert.AreEqual( ps, m.GetTransport() );
            Assert.AreEqual( who, mh.xsender );
            Assert.IsNotNull( mh.xmsg );
            Assert.AreEqual( MyValueFactory.mt_add, mh.xmsg.GetXType );
            Assert.AreEqual( 0, mh.xmsg.Count );

        }

        [Test]
        public void Packet3()
        {
          

            SessionPacket ph = m;

            FlexBuffer buf = new FlexBuffer(new byte[] { (byte)VERSION, 
                unchecked((byte)BYTE), 2, 0, unchecked((byte)NONE) });
            mh.handled = true;

            ph.SessionPacket( who, buf );

            Assert.AreEqual( What.TESTMESSAGEHANDLERMESSAGE, mh.what );
            Assert.AreEqual( ps, m.GetTransport() );
            Assert.AreEqual( who, mh.xsender );
            Assert.IsNotNull( mh.xmsg );
            Assert.AreEqual( MyValueFactory.mt_add_result, mh.xmsg.GetXType );
            Assert.AreEqual( 0, mh.xmsg.Count );

        }

        [Test]
        public void Packet4()
        {
           

            SessionPacket ph = m;

            FlexBuffer buf = new FlexBuffer(new byte[] { (byte)VERSION, 
                unchecked((byte)BYTE), 2, 0, unchecked((byte)NONE) });

            mh.handled = false;

            ph.SessionPacket( who, buf );

            Assert.AreEqual(What.OOB_NOTIFY_HANDLER, mh.what);
            Assert.AreEqual( ps, m.GetTransport() );
            Assert.AreEqual( who, mh.xsender );
            Assert.IsNotNull( mh.xmsg );
            Assert.AreEqual( MyValueFactory.mt_add_result, mh.xmsg.GetXType );
            Assert.AreEqual( 0, mh.xmsg.Count );

        }

        //[Test]
        //public void Down() 
        //{
        //    SetUpBeforeEveryTest();
        //    PacketHandler ph = m;

        //    Assert.IsNull( mh.what );
        //    Assert.IsNull( mh.src );

        //    ph.Down(ps);

        //    Assert.AreEqual( mh.src, m );
        //    Assert.AreEqual( mh.what, What.TESTMESSAGEHANDLERDOWN );
        //    Assert.AreEqual( ps, m.GetPacketSource() );
        //}

        #endregion Test PacketHandler methods

        #region MessagizerMessageSource methods

        //[Test]
        //public void Close1() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    Assert.IsNull( ps.what );

        //    mms.Close( true );

        //    Assert.AreEqual(ps.what, What.PACKETSOURCECLOSE);
        //    Assert.IsTrue( ps.reset );
        //}

        //[Test]
        //public void Close2()
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    Assert.IsNull( ps.what );

        //    mms.Close( false );

        //    Assert.AreEqual( ps.what, What.PACKETSOURCECLOSE );
        //    Assert.IsFalse( ps.reset );
        //}

        //[Test]
        //public void LocalAddress() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    EndPoint isa = new IPEndPoint( IPAddress.Parse( "127.0.0.1" ), 23 );
        //    ps.localAddress  =  isa;

        //    Assert.AreEqual( isa, mms.LocalAddress() );
        //}
    	
        //[Test]
        //public void RemoteAddress() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    EndPoint isa = new IPEndPoint( IPAddress.Parse( "127.0.0.1" ), 23 );
        //    ps.remoteAddress = isa;

        //    Assert.AreEqual( isa, mms.RemoteAddress() );
        //}
    	
        //[Test]
        //public void ShutdownInput() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    Assert.IsNull( ps.what );
        //    m.ShutdownInput();
        //    Assert.AreEqual( ps.what, What.PACKETSOURCESHUTDOWNINPUT );
        //}
    	
        //[Test]
        //public void ShutdownOutput() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    Assert.IsNull( ps.what );
        //    m.ShutdownOutput();

        //    Assert.AreEqual( ps.what, What.PACKETSOURCESHUTDOWNOUTPUT );
        //}
    	
        //[Test]
        //public void Stop() 
        //{
        //    SetUpBeforeEveryTest();

        //    MessagizerMessageSource mms = m;
        //    Assert.IsNull( ps.what );

        //    mms.Stop();
        //    Assert.AreEqual(ps.what, What.PACKETSOURCESTOP);
        //}

   /*     public void Message1()
        {
          
            
            MessagizerMessageSource mms = m;

            Message msg = new Message( MyValueFactory.mt_add, vf );
            ps.headerSize = 0;

            Assert.IsNull( ps.what );

            mms.Messagex( who, msg );

            Assert.AreEqual( What.PACKETSOURCEPACKET, ps.what );
            Assert.AreEqual( who, ps.xrecipient );
            Assert.IsNotNull( ps.xbuf );
            Assert.AreEqual( 4, ps.xbuf.Length );
            Assert.AreEqual( VERSION, ps.xbuf[ 0 ] );
            Assert.AreEqual((byte)1, ps.xbuf[1]);
            Assert.AreEqual((byte)0, ps.xbuf[2]);
            Assert.AreEqual( NONE, ps.xbuf[ 3 ] );
        }

        public void Message2()
        {
          
            MessagizerMessageSource mms = m;

            Message msg = new Message( MyValueFactory.mt_add, vf );
            ps.headerSize = 8;

            Assert.IsNull( ps.what );

            mms.Messagex( who, msg );

            Assert.AreEqual( What.PACKETSOURCEPACKET, ps.what );
            Assert.AreEqual( who, ps.xrecipient );
            Assert.IsNotNull( ps.xbuf );
            Assert.AreEqual( 12, ps.xbuf.Length );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 0 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 1 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 2 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 3 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 4 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 5 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 6 ] );
            Assert.AreEqual( ( byte ) 0, ps.xbuf[ 7 ] );
            Assert.AreEqual( VERSION, ps.xbuf[ 8 ] );
            Assert.AreEqual( ( byte ) 1, ps.xbuf[ 9 ] );
            Assert.AreEqual((byte)0, ps.xbuf[9]);
            Assert.AreEqual( NONE, ps.xbuf[ 11 ] );
        } */

	    
        #endregion MessagizerMessageSource methods

        // Do following test in stress testing.
	    /*
        [Test]
	    public void SendreceiveMessageIterative() 
	    {
            SetUpBeforeEveryTest();
		    report = false;

            int n = 100000;
    		
		    //long t0 = Timer.Init();
            //Timer t0 = new Timer().Init();
		    for (int i = 0; i < n; i++) {
			    MyMessageHandler.Send( m, ps );
			    Assert.IsTrue(ps.checkFlag);
		    }
		    //long t1 = t0.ElapsedNanos();
		    //double t = (t1-t0) / 1000000000.0;
		    //Console.WriteLine( "took " + t + " to send " + n  + " ( " + n/t + " + per sec)\n" );
	    }
        */

        public enum What
        {
            PACKETSOURCEPACKET,
            TESTMESSAGEHANDLERMESSAGE,
            OOB_QUERY_SOURCE,
		    OOB_CONTROL_SOURCE,
		    OOB_NOTIFY_SOURCE,
		    OOB_QUERY_HANDLER,
		    OOB_CONTROL_HANDLER,
		    OOB_NOTIFY_HANDLER
        };

        #region MyPacketSource
        public class MyPacketSource : TransportPacket
        {
            TestMessagizer _tm;

            public What what;
            
            public bool checkFlag = true;
            //public PktSrcMessageHandler mh;
            
            public Messagizer m;

            public SessionPacket handler;
            public int headerSize = 0;
            public bool reset;
            public Who xrecipient;

            public Object xquery;
            /** */
            public Object xquery_result;
            /** */
            public Object xcontrol;
            /** */
            public Object xvalue;
            /** */
            public Object xevent;

            public byte[] xbuf;

            public int HeaderSize()
            {
                return headerSize;
            }

            public MyPacketSource()
            {
                // nothing to do.
            }

            public MyPacketSource( TestMessagizer tm )
            {
                _tm = tm;

            }
            /**
            public class PktSrcMessageHandler : MessageHandler<MessagizerMessageSource>
            {
                MyPacketSource _mps;

                public PktSrcMessageHandler()
                {
                    // nothing to do.
                }

                public PktSrcMessageHandler( MyPacketSource mps )
                {
                    _mps = mps;
                    
                }

                public void Up( MessagizerMessageSource src ) 
                {
                    Console.WriteLine( "up" );
                }

                public bool Message( MessagizerMessageSource src, Who sender, Message msg )
				{
				    if (report)
					    Console.WriteLine( "got reply " + msg );
    				
				    
				    if ( (int)(msg.Get(mf_result)) != 9 ) 
					    _mps.checkFlag = false;

                    return true;
    				
			    }

                public void Down( MessagizerMessageSource src ) 
                {
                    Console.WriteLine( "down" );
                }
            }
    		**/
		    public void Packet( Who recipient, FlexBuffer buf ) 
		    {
                what = What.PACKETSOURCEPACKET;
                this.xrecipient = recipient;
                this.xbuf = buf.GetAvailBytes();
		    }
		    

            #region Transport Members

            public object TransportQuery( object query )
            {
                what = What.OOB_QUERY_SOURCE;
                xquery = query;
                return xquery_result;
            }

            public void TransportControl( object control, object value )
            {
                what = What.OOB_CONTROL_SOURCE;
                xcontrol = control;
                xvalue = value;
            }

            public void TransportNotify( object eventObj )
            {
                what = What.OOB_NOTIFY_SOURCE;
                xevent = eventObj;
            }

            #endregion

           
            #region TransportPacket Members


            public void TransportPacket(Who recipient, FlexBuffer buf)
            {
                what = What.PACKETSOURCEPACKET;
                this.xrecipient = recipient;
                this.xbuf = buf.GetAvailBytes();
            }

            #endregion

            #region Transport<SessionPacket> Members


            public void SetSession(SessionPacket session)
            {
                this.handler = session;
            }

            #endregion

            #region Transport<SessionPacket> Members


            public SessionPacket GetSession()
            {
                return handler;
            }

            #endregion
        }

        #endregion MyPacketSource

        #region MyMessageHandler

        public class MyMessageHandler : SessionMessage
        {
            public Enum what;
            public Who xsender;
            public Message xmsg;
            public bool handled;

            public Object xquery;
            public Object xquery_result;
            public Object xcontrol;
            public Object xvalue;
            public Object xevent;

            

            public bool Message( Who sender, Message msg ) 
		    {

                what = What.TESTMESSAGEHANDLERMESSAGE;
                this.xsender = sender;
                this.xmsg = msg;
                return handled;

                /**
			    if (report)
		    	    Console.WriteLine( "message: "+msg );
    		    
		        if (msg.IsType( mt_add ))
		        {
		    	    int? x = (int?)( msg.Get( mf_x ) );
                    int? y = ( int? ) ( msg.Get( mf_y ) );

                    Int32 r = ( int ) ( x + y );
    		    	
		    	    Message rmsg = msg.Reply( mt_add_result );
		    	    rmsg.Add( mf_result, r );
    		    	
		    	    if (report)
		    		    Console.WriteLine( "replying with "+rmsg );
    		    	
		    	    src.Message( sender, rmsg );
		        }

                return true;
                 * */
	        }

         

            #region Session Members

            public object SessionQuery( object query )
            {
                what = What.OOB_QUERY_HANDLER;
                xquery = query;
                return xquery_result;
            }

            public void SessionControl( object control, object value )
            {
                what = What.OOB_CONTROL_HANDLER;
                xcontrol = control;
                xvalue = value;
            }

            public void SessionNotify( object eventObj )
            {
                what = What.OOB_NOTIFY_HANDLER;
                xevent = eventObj;
            }

            #endregion

            #region SessionMessage Members

            public bool SessionMessage(Who sender, Message msg)
            {
                what = What.TESTMESSAGEHANDLERMESSAGE;
                this.xsender = sender;
                this.xmsg = msg;
                return handled;
            }

            #endregion
        }
        #endregion MyMessageHandler

        public class MyValueFactory : DefaultValueFactory
        {
            private readonly static TypeMap types = new TypeMap();

           
            private readonly static Class2TypeMap class2type = 
                new Class2TypeMap();

            public readonly static XType mt_add = types.Add( new XType( 1, "add" ) );

            public readonly static XType mt_add_result = types.Add( new XType( 2, "add_result" ) );

            public static readonly Field mf_x = new Field(3, "x");

            public static readonly Field mf_y = new Field(4, "y"); 

            public readonly static Field mf_result =  new Field( 5, "xresult");

            static MyValueFactory()
            {
                DefaultValueFactory.Init( types,  class2type );

                mt_add.PutValidator( mf_x, Validator_int.Get( 0 ) );
                mt_add.PutValidator( mf_y, Validator_int.Get( 0 ) );
                mt_add.PutValidator( DefaultValueFactory._mf__messageId, Validator_long.Get( 0 ) );
                
                mt_add_result.PutValidator( mf_result, Validator_int.Get( 0 ) );
                mt_add_result.PutValidator( DefaultValueFactory._mf__messageId, Validator_long.Get( 0 ) );
                mt_add_result.PutValidator( DefaultValueFactory._mf__inReplyTo, Validator_long.Get( 0 ) );

            }

            public MyValueFactory() : base("none:", types, class2type)
            {
                // nothing.
            }
        }
    }
}

File: binding-java/runtime/src/test/java/org/apache/etch/bindings/java/transport/TestMessagizer.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.transport;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;

import org.apache.etch.bindings.java.msg.Field;
import org.apache.etch.bindings.java.msg.Message;
import org.apache.etch.bindings.java.msg.Type;
import org.apache.etch.bindings.java.msg.TypeMap;
import org.apache.etch.bindings.java.msg.ValueFactory;
import org.apache.etch.bindings.java.support.Class2TypeMap;
import org.apache.etch.bindings.java.support.DefaultValueFactory;
import org.apache.etch.bindings.java.support.Validator_int;
import org.apache.etch.bindings.java.support.Validator_long;
import org.apache.etch.util.FlexBuffer;
import org.apache.etch.util.Resources;
import org.apache.etch.util.core.Who;
import org.apache.etch.util.core.io.SessionPacket;
import org.apache.etch.util.core.io.Transport;
import org.apache.etch.util.core.io.TransportPacket;
import org.junit.Test;


/** */
public class TestMessagizer
{
	private final static byte VERSION = 3; // BinaryTaggedData.VERSION;
	private final static byte NONE = -127; // TypeCode.NONE;
	
	/** */
	public TestMessagizer()
	{
		transport = new MyTransportPacket();
		session = new MySessionMessage();
		
		Resources res = new Resources();
		res.put( Transport.VALUE_FACTORY, vf );
		
		m = new Messagizer( transport, "foo:?Messagizer.format=binary", res );
		m.setSession( session );
	}

	private final ValueFactory vf = new MyValueFactory( "tcp:" );

	private final MyTransportPacket transport;

	private final MySessionMessage session;

	private final Messagizer m;
	
	/** @throws Exception */
	@Test
	public void construct1() throws Exception
	{
		assertSame( transport, m.getTransport() );
		assertSame( session, m.getSession() );
		assertSame( m, transport.getSession() );
	}

	////////////////////////////////
	// Test PacketHandler methods //
	////////////////////////////////

//	/** @throws Exception */
//	@Test
//	public void up() throws Exception
//	{
//		PacketHandler ph = m;
//
//		assertNull( mh.what );
//		assertNull( mh.src );
//
//		ph.up( ps );
//
//		assertSame( What.TESTMESSAGEHANDLERUP, mh.what );
//		assertSame( m, mh.src );
//		assertSame( ps, m.getSource() );
//	}
	
	private final Who who = new Who() { /* nothing */ };

	/** @throws Exception */
	@Test
	public void packet1() throws Exception
	{
		FlexBuffer buf = new FlexBuffer( new byte[]
		{ VERSION, 1, 0, NONE } );
		
		session.msg_handled = true;
		
		m.sessionPacket( who, buf );
		
		assertSame( What.SESSION_MESSAGE, session.what );
		assertSame( who, session.sender );
		assertSame( MyValueFactory.mt_add, session.msg.type() );
		assertEquals( 0, session.msg.size() );
		assertNull( session.event );
	}

	/** @throws Exception */
	@Test
	public void packet2() throws Exception
	{
		FlexBuffer buf = new FlexBuffer( new byte[]
		{ VERSION, 1, 0, NONE } );
		
		session.msg_handled = false;
		
		m.sessionPacket( who, buf );
		
		assertSame( What.SESSION_NOTIFY, session.what );
		assertSame( who, session.sender );
		assertSame( MyValueFactory.mt_add, session.msg.type() );
		assertEquals( 0, session.msg.size() );
		assertNotNull( session.event );
		
		UnwantedMessage um = (UnwantedMessage) session.event;
		assertSame( session.sender, um.sender );
		assertSame( session.msg, um.msg );
	}

	/** @throws Exception */
	@Test
	public void packet3() throws Exception
	{
		FlexBuffer buf = new FlexBuffer( new byte[]
		{ VERSION, 2, 0, NONE } );
		
		session.msg_handled = true;
		
		m.sessionPacket( who, buf );
		
		assertSame( What.SESSION_MESSAGE, session.what );
		assertSame( who, session.sender );
		assertNotNull( session.msg );
		assertSame( MyValueFactory.mt_add_result, session.msg.type() );
		assertEquals( 0, session.msg.size() );
		assertNull( session.event );
	}

	/** @throws Exception */
	@Test
	public void packet4() throws Exception
	{
		FlexBuffer buf = new FlexBuffer( new byte[]
		{ VERSION, 2, 0, NONE } );
		
		session.msg_handled = false;
		
		m.sessionPacket( who, buf );
		
		assertSame( What.SESSION_NOTIFY, session.what );
		assertSame( who, session.sender );
		assertSame( MyValueFactory.mt_add_result, session.msg.type() );
		assertEquals( 0, session.msg.size() );
		assertNotNull( session.event );
		
		UnwantedMessage um = (UnwantedMessage) session.event;
		assertSame( session.sender, um.sender );
		assertSame( session.msg, um.msg );
	}

	/** @throws Exception */
	@Test
	public void message1() throws Exception
	{
		Message msg = new Message( MyValueFactory.mt_add, vf );
		transport.header_size = 0;
		
		assertNull( transport.what );
		
		m.transportMessage( who, msg );

		assertSame( What.TRANSPORT_PACKET, transport.what );
		assertSame( who, transport.recipient );
		assertNotNull( transport.buf );
		assertEquals( 4, transport.buf.length );
		assertEquals( VERSION, transport.buf[0] ); // version
		assertEquals( (byte) 1, transport.buf[1] ); // message type
		assertEquals( (byte) 0, transport.buf[2] ); // length
		assertEquals( NONE, transport.buf[3] ); // none
	}

	/** @throws Exception */
	@Test
	public void message2() throws Exception
	{
		Message msg = new Message( MyValueFactory.mt_add_result, vf );
		transport.header_size = 8;
		
		assertNull( transport.what );
		
		m.transportMessage( who, msg );

		assertSame( What.TRANSPORT_PACKET, transport.what );
		assertSame( who, transport.recipient );
		assertNotNull( transport.buf );
		assertEquals( 12, transport.buf.length );
		assertEquals( (byte) 0, transport.buf[0] ); // packet header...
		assertEquals( (byte) 0, transport.buf[1] );
		assertEquals( (byte) 0, transport.buf[2] );
		assertEquals( (byte) 0, transport.buf[3] );
		assertEquals( (byte) 0, transport.buf[4] );
		assertEquals( (byte) 0, transport.buf[5] );
		assertEquals( (byte) 0, transport.buf[6] );
		assertEquals( (byte) 0, transport.buf[7] );
		assertEquals( VERSION, transport.buf[8] ); // version
		assertEquals( (byte) 2, transport.buf[9] ); // message type
		assertEquals( (byte) 0, transport.buf[10] ); // length
		assertEquals( NONE, transport.buf[11] ); // none
	}

	/** @throws Exception */
	@Test
	public void sessionQuery() throws Exception
	{
		Object QUERY = "foo";
		Object RESULT = "bar";
		
		assertNull( session.what );
		assertNull( session.query );
		session.query_result = RESULT;
		
		Object result = m.sessionQuery( QUERY );

		assertSame( What.SESSION_QUERY, session.what );
		assertSame( QUERY, session.query );
		assertSame( RESULT, result );
	}

	/** @throws Exception */
	@Test
	public void sessionControl() throws Exception
	{
		Object CONTROL = "foo";
		Object VALUE = "bar";
		
		assertNull( session.what );
		assertNull( session.control );
		assertNull( session.value );
		
		m.sessionControl( CONTROL, VALUE );

		assertSame( What.SESSION_CONTROL, session.what );
		assertSame( CONTROL, session.control );
		assertSame( VALUE, session.value );
	}

	/** @throws Exception */
	@Test
	public void sessionNotify() throws Exception
	{
		Object EVENT = "foo";
		
		assertNull( session.what );
		assertNull( session.event );
		
		m.sessionNotify( EVENT );

		assertSame( What.SESSION_NOTIFY, session.what );
		assertSame( EVENT, session.event );
	}

	/** @throws Exception */
	@Test
	public void transportQuery() throws Exception
	{
		Object QUERY = "foo";
		Object RESULT = "bar";
		
		assertNull( transport.what );
		assertNull( transport.query );
		transport.query_result = RESULT;
		
		Object result = m.transportQuery( QUERY );

		assertSame( What.TRANSPORT_QUERY, transport.what );
		assertSame( QUERY, transport.query );
		assertSame( RESULT, result );
	}

	/** @throws Exception */
	@Test
	public void transportControl() throws Exception
	{
		Object CONTROL = "foo";
		Object VALUE = "bar";
		
		assertNull( transport.what );
		assertNull( transport.control );
		assertNull( transport.value );
		
		m.transportControl( CONTROL, VALUE );

		assertSame( What.TRANSPORT_CONTROL, transport.what );
		assertSame( CONTROL, transport.control );
		assertSame( VALUE, transport.value );
	}

	/** @throws Exception */
	@Test
	public void transportNotify() throws Exception
	{
		Object EVENT = "foo";
		
		assertNull( transport.what );
		assertNull( transport.event );
		
		m.transportNotify( EVENT );

		assertSame( What.TRANSPORT_NOTIFY, transport.what );
		assertSame( EVENT, transport.event );
	}

	/**
	 * Reports activities which occur.
	 */
	public enum What
	{
		/** */ TRANSPORT_PACKET,
		/** */ TRANSPORT_QUERY,
		/** */ TRANSPORT_CONTROL,
		/** */ TRANSPORT_NOTIFY,
		/** */ SESSION_MESSAGE,
		/** */ SESSION_QUERY,
		/** */ SESSION_CONTROL,
		/** */ SESSION_NOTIFY
	};

	/** */
	public static class MyTransportPacket implements TransportPacket
	{
		/** */ public int header_size;
		/** */ public What what;
		/** */ public Who recipient;
		/** */ public byte[] buf;
		/** */ public Object query;
		/** */ public Object query_result;
		/** */ public Object control;
		/** */ public Object value;
		/** */ public Object event;
		
		public int headerSize()
        {
	        return header_size;
        }

		public void transportPacket( Who recipient, FlexBuffer buf ) throws Exception
        {
	        what = What.TRANSPORT_PACKET;
	        this.recipient = recipient;
	        this.buf = buf.getAvailBytes();
        }

		public Object transportQuery( Object query ) throws Exception
		{
			what = What.TRANSPORT_QUERY;
			this.query = query;
			return query_result;
		}

		public void transportControl( Object control, Object value ) throws Exception
		{
			what = What.TRANSPORT_CONTROL;
			this.control = control;
			this.value = value;
		}

		public void transportNotify( Object event ) throws Exception
		{
			what = What.TRANSPORT_NOTIFY;
			this.event = event;
		}

		public SessionPacket getSession()
		{
			return session;
		}

		public void setSession( SessionPacket session )
		{
			this.session = session;
		}
		
		private SessionPacket session;
	}

	/** */
	public static class MySessionMessage implements SessionMessage
	{
		/** */
		public What what;
		/** */
		public Who sender;
		/** */
		public Message msg;
		/** */
		public boolean msg_handled;
		/** */
		public Object query;
		/** */
		public Object query_result;
		/** */
		public Object control;
		/** */
		public Object value;
		/** */
		public Object event;

		public boolean sessionMessage( Who sender, Message msg ) throws Exception
		{
			what = What.SESSION_MESSAGE;
			this.sender = sender;
			this.msg = msg;
			return msg_handled;
		}

		public Object sessionQuery( Object query )
		{
			what = What.SESSION_QUERY;
			this.query = query;
			return query_result;
		}

		public void sessionControl( Object control, Object value )
		{
			what = What.SESSION_CONTROL;
			this.control = control;
			this.value = value;
		}

		public void sessionNotify( Object event )
		{
			what = What.SESSION_NOTIFY;
			this.event = event;
		}
	}
	
	/** Private value factory for testing. */
	public static class MyValueFactory extends DefaultValueFactory
	{
		/**
		 * Construct MyValueFactory.
		 * @param uri 
		 */
		public MyValueFactory( String uri )
		{
			super( uri, types, class2type );
		}
		
		private final static TypeMap types = new TypeMap();
		
		private final static Class2TypeMap class2type = new Class2TypeMap();

		static
		{
			DefaultValueFactory.init( types, class2type );
		}

		/** */
		public final static Type mt_add = types.add( new Type( 1, "add" ) );

		/** */
		public final static Type mt_add_result = types.add( new Type( 2, "add_result" ) );

		/** */
		public final static Field mf_x = new Field( 3, "x" );

		/** */
		public final static Field mf_y = new Field( 4, "y" );

		/** */
		public final static Field mf_result = new Field( 5, "xresult" );
		
		static
		{
			mt_add.putValidator( mf_x, Validator_int.get( 0 ) );
			mt_add.putValidator( mf_y, Validator_int.get( 0 ) );
			mt_add.putValidator( _mf__messageId, Validator_long.get( 0 ) );

			mt_add_result.putValidator( mf_result, Validator_int.get( 0 ) );
			mt_add_result.putValidator( _mf__messageId, Validator_long.get( 0 ) );
			mt_add_result.putValidator( _mf__inReplyTo, Validator_long.get( 0 ) );
		}
	}
}
